backend/abstract/selenium_scraper.py

```diff
@@ -0,0 +1,208 @@
+                    'page_title': page_title,
+                    'page_source': page_source,
+                    'detected_404': detected_404
+                    }
+        else:
+            return False
+
+    def collect_links(self):
+        """
+
+        """
+        if self.driver is None:
+            raise ProcessorException('Selenium Drive not yet started: Cannot collect links')
+
+        elems = self.driver.find_elements_by_xpath("//a[@href]")
+        return [elem.get_attribute("href") for elem in elems]
+
+    def search(self, query):
+        """
+        Search for items matching the given query
+
+        The real work is done by the get_items() method of the descending
+        class. This method just provides some scaffolding and post-processing
+        of results via `after_search()`, if it is defined.
+
+        :param dict query:  Query parameters
+        :return:  Iterable of matching items, or None if there are no results.
+        """
+        self.start_selenium()
+        # Returns to default position; i.e., 'data:,'
+        self.reset_current_page()
+        # Sets timeout to 60
+        self.set_page_load_timeout()
+
+        # Normal Search function to be used To be implemented by descending classes!
+        try:
+            posts = self.get_items(query)
+        except Exception as e:
+            # Ensure Selenium always quits
+            self.quit_selenium()
+            raise e
+
+        if not posts:
+            return None
+
+        # search workers may define an 'after_search' hook that is called after
+        # the query is first completed
+        if hasattr(self, "after_search") and callable(self.after_search):
+            posts = self.after_search(posts)
+
+        return posts
+
+    def start_selenium(self):
+        """
+        Start a headless browser
+        """
+        options = Options()
+        options.headless = True
+        options.add_argument('--headless')
+        options.add_argument('--no-sandbox')
+        options.add_argument('--disable-dev-shm-usage')
+
+        try:
+            self.driver = webdriver.Chrome(options=options)
+        except (SessionNotCreatedException, WebDriverException) as e:
+            if "binary not found" in str(e):
+                raise ProcessorException("Chromium binary is not available.")
+            if "only supports Chrome" in str(e):
+                raise ProcessorException("Your chromedriver version is incompatible with your Chromium version:\n  (%s)" % e)
+            else:
+                raise ProcessorException("Could not connect to Chromium (%s)." % e)
+
+    def quit_selenium(self):
+        """
+        Always attempt to close the browser otherwise multiple versions of Chrome will be left running.
+
+        And Chrome is a memory hungry monster.
+        """
+        try:
+            self.driver.quit()
+        except:
+            pass
+
+    def after_search_completed(self):
```

Comment:
I actually tried that originally and that was the idea for this method. `BasicProcessor.after_process()` only runs if `process()` completes successfully. I need to ensure Chrome and the webdriver _always_ close, even if it is interrupted or if there is an error. Otherwise you're left with memory hungry Chrome hanging out and, worse off, if another datasource uses Selenium, it will start new instances of both.

But looking at it a bit more, I could override both `after_process` and `remove_files`. I think that will catch `process` success as well as any errors if you don't like the `after_search_completed` method being added.

Reply:
Actually, it might make sense to move this `after_search_completed` to a `clean_up` method in `BasicProcessor` that always runs after `process` whether it completes successfully or not. I did it in search since I was thinking there might be other things I would like to do after a search was completed.

It could be where we remove temp files for example (which currently happens in `after_process` and `remove_files` which is in abort and runs on its own if there is an exception). Though if we want to be able to resume datasources, we have to change that anyway since we cannot remove those temp files.

Reply:
I really like the former idea - a general-purpose way to add code that will *always* be executed after a processor finishes or crashes would likely be useful in more cases. That's worth adding an extra 'hook'. And as you say it would solve the problem of temporary files being left behind. Resuming datasources is probably best added as a feature in one go, so we don't need to accomodate it yet.

Reply:
Ok, added a `clean_up` method to BasicWorker that runs after `work` whether it is successful or not. BasicProcessor then uses that and removes any staging areas. Selenium processor then adds to that method for quit.

I could have added clean_up in the BasicProcessor, but it would actually need to run in each except block and just seemed a bit messier. Plus now any worker can use it.